### ISSUE TITLE
Datatable should paginate by its own

### ISSUE THREAD

**Issue by sinedie** [carbon-design-system/carbon-components-svelte#702]
Hey there. Could be good to enable automatic pagination on datatable component.

Eg. I have a table with 1000 rows, but I want to see 10 at the same time and be able to sort it. I can create the pagination and handle which rows are visible and not, but when sorting it just sort the 10 rows that are visible, not the entire dataset.

### Possible solution
Create an Integer `visibleRows` prop on the datatable component and just render the number of row passed.

**Comment by metonym** [comment]
Actually, I'm inclined to agree.

Leaving pagination to the consumer was meant to account for large datasets (i.e., 100k+ rows) where client-side sorting would be slow.

However, I've also run into scenarios where I have <500 rows; that's when client-side sorting would be more feasible.

**Comment by brunnerh** [comment]
@metonym Pagination is not a great user experience, especially on mobile devices. I would prefer keeping the scroll behavior and enable virtualization for large data sets. Right now i have to use a different component like [revolist/svelte-datagrid](https://github.com/revolist/svelte-datagrid) to handle this.

As large data sets are fairly common, it would be appreciated if virtualization could be handled by the Carbon data grid as well.

**Comment by malinowskip** [comment]
Built-in pagination would be a very useful feature – or at least the ability to provide a completely custom sorting function to each column. Right now, it’s pretty straightforward to implement pagination with some additional business logic and the `Pagination` component. However, sorting itself is still limited to the data that’s actually loaded into the datatable.

Maybe the datatable could receive something like an additional `sortOrder` argument – specifying a column index and sorting order, e.g. `[0, "desc"]`. Then the consumer would simply subscribe to the `on:click:header` event, swap out the rows (which are already reactive) as needed, and update the `sortOrder` prop.

**Comment by metonym** [comment]
For anyone interested in a sample implementation, take a look at [repos.svelte](https://github.com/carbon-design-system/carbon-tutorial-svelte/blob/main/src/routes/repos.svelte) from the [Svelte tutorial](https://github.com/carbon-design-system/carbon-tutorial-svelte).

---

In addition, there is an unfinished proof of concept for a re-usable [`PaginatedDatatable` component](https://github.com/carbon-design-system/carbon-components-svelte/blob/paginated-datatable/src/PaginatedDataTable/PaginatedDataTable.svelte).

**Comment by malinowskip** [comment]
Thanks for pointing to the POC. One issue that I see is that data is sorted twice – once by the `DataTable` component, and then by `PaginatedDataTable`. When you click on a sortable column in `PaginatedDataTable`, what I think happens is:

1. DataTable` dispatches a `click:header` event;
2. `PaginatedDataTable` responds to the event by performing its sort;
2. `DataTable` simultaneously applies its (redundant) sort to the visible rows.

 Do you think there’s a way to suppress `DataTable`’s sort without breaking its API?

**Comment by pwwang** [comment]
The solution should be simple since we have already had sorting.

We just need to calculate which items to show on the current page. Say we have a function to get those items from all rows, page size, and current page:

```js

paginate = function ( items, pageSize, page) {
    return items
      .slice(
        (page- 1) * pageSize,
        (page- 1) * pageSize + pageSize
      )
}

```

Then, calculate the items, reactively:

```js
$: pagedRows = sorting
        ? paginate(sortedRows, pageSize, page)
        : paginate(rows, pageSize, page);
```

Now replace the rows to render the table body:

(in `DataTable.svelte`)
```jsx
    <TableBody>
      {#each sorting ? sortedRows : rows as row, i (row.id)}
```
with:
```jsx
    <TableBody>
      {#each pagedRows as row, i (row.id)}
```

That's it.

We just got trapped by the idea of reusing `DataTable`, but it's time to change `DataTable` to adopt pagination, by maybe adding some props to enable pagination.

**Comment by malinowskip** [comment]
That’s great. Here’s a rough implementation: https://github.com/malinowskip/carbon-components-svelte/commit/4fddc15fa44f426becb4fef0fb6c2f7386c5e5c4.

**Comment by pwwang** [comment]
@malinowskip You may also need to replace the `rows` in selection, expansion, etc.

**Comment by malinowskip** [comment]
I’ve just looked at all the examples in the docs and they seem to be working fine. Selection/expansion uses rows’ `id`’s, so it’s independent from pagination. Do you mean that selection/expansion state should be maintained only for rows on the current page?

**Comment by pwwang** [comment]
No, then that should be fine. Great work!

**Comment by metonym** [comment]
@kwieszalka-maystreet added support for a paginated `DataTable` component, released in [v0.54.0](https://github.com/carbon-design-system/carbon-components-svelte/releases/tag/v0.54.0).

The approach is very elegant. All you need to do is bind the `pageSize`, `page` props from `Pagination` and pass them to the `DataTable`.

```svelte
<script>
  import { DataTable, Pagination } from "carbon-components-svelte";

  const pagination = { pageSize: 5, page: 1 };
</script>

<DataTable
  sortable
  title="Load balancers"
  description="Your organization's active load balancers."
  headers={[
    { key: "name", value: "Name" },
    { key: "protocol", value: "Protocol" },
    { key: "port", value: "Port" },
    { key: "cost", value: "Cost", display: (cost) => cost + " €" },
    {
      key: "expireDate",
      value: "Expire date",
      display: (date) => new Date(date).toLocaleString(),
      sort: (a, b) => new Date(a) - new Date(b),
    },
  ]}
  pageSize={pagination.pageSize}
  page={pagination.page}
  rows={[
    {
      id: "a",
      name: "Load Balancer 3",
      protocol: "HTTP",
      port: 3000,
      cost: 100,
      expireDate: "2020-10-21",
    },
    {
      id: "b",
      name: "Load Balancer 1",
      protocol: "HTTP",
      port: 443,
      cost: 200,
      expireDate: "2020-09-10",
    },
    {
      id: "c",
      name: "Load Balancer 2",
      protocol: "HTTP",
      port: 80,
      cost: 150,
      expireDate: "2020-11-24",
    },
    {
      id: "d",
      name: "Load Balancer 6",
      protocol: "HTTP",
      port: 3000,
      cost: 250,
      expireDate: "2020-12-01",
    },
    {
      id: "e",
      name: "Load Balancer 4",
      protocol: "HTTP",
      port: 443,
      cost: 550,
      expireDate: "2021-03-21",
    },
    {
      id: "f",
      name: "Load Balancer 5",
      protocol: "HTTP",
      port: 80,
      cost: 400,
      expireDate: "2020-11-14",
    },
  ]}
/>
<Pagination
  bind:pageSize={pagination.pageSize}
  bind:page={pagination.page}
  totalItems={6}
  pageSizeInputDisabled
/>

```

[Example](https://carbon-components-svelte.onrender.com/components/DataTable#sortable-with-pagination)
